### PR TITLE
make realnvp and nsf layers as part of the pkg

### PR DESCRIPTION
**Pull request by zuhengxu**
As discussed in #36 (see https://github.com/TuringLang/NormalizingFlows.jl/issues/36#issuecomment-3034164318), I'm moving the `AffineCoupling` and `NeuralSplineLayer` from the example to `src/` so it can be called. 

- [x] move `AffineCoupling` and `NeuralSplineLayer` into `src`
- [x] create a out-of-box `realnvp` and a `neuralsplineflow` constructor. For the `realnvp`, follow the default architecture as mentioned in [Advances in Black-Box VI: Normalizing Flows, Importance Weighting, and Optimization](https://proceedings.neurips.cc/paper/2020/file/c91e3483cf4f90057d02aa492d2b25b1-Paper.pdf).
- [x] add test for the two flows
- [x] documentation

### CODE REVIEW

**Comment by zuhengxu** [comment]
After the recent Mooncake update (`Mooncake v1.4.140`, `DI v), all realnvp and neural spline flow tests/demos failed. After spending some time debugging it, it seems like its due to the failure of `Mooncake` traversing through nested structures? Here is a minimal example: 
```julia
using Random, Distributions, LinearAlgebra
using Bijectors
using Bijectors: partition, combine, PartitionMask
using Mooncake, Enzyme, ADTypes
import DifferentiationInterface as DI

# just define a MLP 
function mlp3(
    input_dim::Int, 
    hidden_dims::Int, 
    output_dim::Int; 
    activation=Flux.leakyrelu,
    paramtype::Type{T} = Float64
) where {T<:AbstractFloat}
    m = Chain(
        Flux.Dense(input_dim, hidden_dims, activation),
        Flux.Dense(hidden_dims, hidden_dims, activation),
        Flux.Dense(hidden_dims, output_dim),
    )
    return Flux._paramtype(paramtype, m)
end


inputdim = 4
mask_idx = 1:2:inputdim
# creat a masking layer
mask = PartitionMask(inputdim, mask_idx)
cdim = length(mask_idx)

x = randn(inputdim)

t_net = mlp3(cdim, 16, cdim; paramtype = Float64)
ps, st = Optimisers.destructure(t_net)
```

the following code runs perfectly
```julia
function loss(ps, st, x, mask)
    t_net = st(ps)
    x₁, x₂, x₃ = partition(mask, x)
    y₁ = x₁ .+ t_net(x₂)
    y = combine(mask, y₁, x₂, x₃)
    # println("y = ", y)
    return sum(abs2, y)
end


loss(ps, st, x, mask)  # return 3.0167880799441793

val, grad = DI.value_and_gradient(
    ls_msk, 
    ADTypes.AutoMooncake(; config = Mooncake.Config()), 
    ps, DI.Cache(st), DI.Constant(x), DI.Constant(mask)
)
```
autograd fails if I wrap `t_net` and `mask` into a struct 
```julia
struct ACL
    mask::Bijectors.PartitionMask
    t::Flux.Chain
end
@functor ACL (t, )

acl = ACL(mask, t_net)
psacl, stacl = Optimisers.destructure(acl)

function loss_acl(ps, st, x)
    acl = st(ps)
    t_net = acl.t
    mask = acl.mask
    x₁, x₂, x₃ = partition(mask, x)
    y₁ = x₁ .+ t_net(x₂)
    y = combine(mask, y₁, x₂, x₃)
    return sum(abs2, y)
end
loss_acl(psacl, stacl, x) # return 3.0167880799441793

val, grad = DI.value_and_gradient(
    loss_acl, 
    ADTypes.AutoMooncake(; config = Mooncake.Config()), 
    psacl, DI.Cache(stacl), DI.Constant(x)
)
```
with error message
```
Unreachable reached at 0x7597f9780297

[1530713] signal 4 (2): Illegal instruction
in expression starting at REPL[47]:1
RRuleZeroWrapper at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:302 [inlined]
|> at ./operators.jl:926 [inlined]
opaque closure at ./<missing>:0
DerivedRule at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:966
unknown function (ip: 0x7597f9780f51)
DynamicDerivedRule at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:1739
RRuleZeroWrapper at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:302
unknown function (ip: 0x7597f977b8ba)
_Trainable_biwalk at /home/zuhdav/.julia/packages/Optimisers/W5seC/src/destructure.jl:108 [inlined]
opaque closure at ./<missing>:0
DerivedRule at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:966 [inlined]
_build_rule! at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:1827
LazyDerivedRule at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:1822 [inlined]
RRuleZeroWrapper at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:302 [inlined]
loss_acl at ./REPL[43]:2 [inlined]
opaque closure at ./<missing>:0
DerivedRule at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interpreter/s2s_reverse_mode_ad.jl:966
unknown function (ip: 0x7597fa9be0df)
jl_apply at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
do_apply at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/builtins.c:831
#prepare_gradient_cache#781 at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interface.jl:509
prepare_gradient_cache at /home/zuhdav/.julia/packages/Mooncake/qKNoG/src/interface.jl:506 [inlined]
prepare_gradient_nokwarg at /home/zuhdav/.julia/packages/DifferentiationInterface/a7NWj/ext/DifferentiationInterfaceMooncakeExt/onearg.jl:114
value_and_gradient at /home/zuhdav/.julia/packages/DifferentiationInterface/a7NWj/src/first_order/gradient.jl:36
unknown function (ip: 0x7597fa9ade93)
jl_apply at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
do_call at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/interpreter.c:126
eval_value at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/interpreter.c:223
eval_stmt_value at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/interpreter.c:174 [inlined]
eval_body at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/interpreter.c:666
jl_interpret_toplevel_thunk at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/interpreter.c:824
jl_toplevel_eval_flex at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/toplevel.c:943
jl_toplevel_eval_flex at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
jl_toplevel_eval_flex at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
jl_toplevel_eval_flex at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
ijl_toplevel_eval_in at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/toplevel.c:994
eval at ./boot.jl:430 [inlined]
eval_user_input at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/REPL/src/REPL.jl:261
repl_backend_loop at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/REPL/src/REPL.jl:368
#start_repl_backend#59 at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/REPL/src/REPL.jl:343
start_repl_backend at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/REPL/src/REPL.jl:340
#run_repl#76 at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/REPL/src/REPL.jl:500
run_repl at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/REPL/src/REPL.jl:486
jfptr_run_repl_10230.1 at /home/zuhdav/.julia/juliaup/julia-1.11.6+0.x64.linux.gnu/share/julia/compiled/v1.11/REPL/u0gqU_QBeOa.so (unknown line)
#1150 at ./client.jl:446
jfptr_YY.1150_14947.1 at /home/zuhdav/.julia/juliaup/julia-1.11.6+0.x64.linux.gnu/share/julia/compiled/v1.11/REPL/u0gqU_QBeOa.so (unknown line)
jl_apply at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
jl_f__call_latest at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/builtins.c:875
#invokelatest#2 at ./essentials.jl:1055 [inlined]
invokelatest at ./essentials.jl:1052 [inlined]
run_main_repl at ./client.jl:430
repl_main at ./client.jl:567 [inlined]
_start at ./client.jl:541
jfptr__start_73597.1 at /home/zuhdav/.julia/juliaup/julia-1.11.6+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
jl_apply at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
true_main at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/jlapi.c:900
jl_repl_entrypoint at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/src/jlapi.c:1059
main at /cache/build/tester-amdci4-12/julialang/julia-release-1-dot-11/cli/loader_exe.c:58
unknown function (ip: 0x759918229d8f)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
Allocations: 435196342 (Pool: 435185964; Big: 10378); GC: 172
[1]    1530713 illegal hardware instruction (core dumped)  julia --project=@. --threads=20
```
`Enzyme` works on this example
```julia
val, grad = DI.value_and_gradient(
    loss_acl, 
    ADTypes.AutoEnzyme(;
            mode=Enzyme.set_runtime_activity(Enzyme.Reverse),
            function_annotation=Enzyme.Const,
        ),
    psacl, DI.Cache(stacl), DI.Constant(x)
)
```
with output
```
(5.453049189044009, [0.00547994735278205, -0.0026771692066405567, -0.005708182026795597, -0.7924439468724516, -0.7544730762364316, -0.16686686701374107, 0.4478351144536696, -0.006762348018489824, 0.004773511138093715, -0.0005490333340741215  …  0.0020767492481565375, 0.00012238617695840655, 0.0028390198598604306, 0.0001673080114345703, -0.28502936510323545, -0.016797239410021175, -0.5721215941140413, -0.0337160467115233, -4.572356477535249, -0.2694563291515709])
```
Any thoughts on this @yebai @willtebbutt?

**Comment by zuhengxu** [comment]
Ah looks like it only has issue when the part of the fields in the structure is annotated by @functor; the following example when the struct only contains a single trainable works:
```julia 
struct Holder
    t::Flux.Chain
end
@functor Holder 

psh, sth = Optimisers.destructure(Holder(t_net))

function loss2(ps, st, x)
    holder = st(ps)
    t_net = holder.t
    y = x .+ t_net(x)
    return sum(abs2, y)
end
loss2(psh, sth, x)  # return 7.408352005690478

val, grad = DI.value_and_gradient(
    loss2, 
    ADTypes.AutoMooncake(; config = Mooncake.Config()), 
    psh, DI.Cache(sth), DI.Constant(x)
)
```
with outputs
```
(7.408352005690478, [0.016964275123278565, -1.6770328865228394, 0.012168583592473805, -1.899494171602285, -0.03156295180782276, 1.1263707232774889, 0.041515954305201684, -1.633822068384897, 0.0291423962210742, -0.06883290364310875  …  -0.5149587512518972, 0.3813047278162709, 0.03035515401339805, 0.004346607224947255, 0.004533179126481482, -0.0033566234747215667, -5.299778474492951, -0.7588844845815451, -0.7914585162356058, 0.5860408690548988])
```

**Comment by yebai** [comment]
@zuhengxu, can you help bisect which Mooncake version / Julia version introduced this bug?

**Comment by zuhengxu** [comment]
> @zuhengxu, can you help bisect which Mooncake version / Julia version introduced this bug?

gGod point! I'll look at this today.

**Comment by zuhengxu** [comment]
> @zuhengxu, can you help bisect which Mooncake version / Julia version introduced this bug?

@yebai I finally nailed down the versions. I believe the bug was introduced by `Mooncake v0.4.124`---tests all pass under `julia v1.11.6`  + `Mooncake v0.4.123` but failed under `julia v1.11.6`  + `Mooncake v0.4.124`.

**Comment by yebai** [comment]
It appears that the remaining issues with Mooncake are minor, likely due to a lack of a specific rule. 

@sunxd3, can you help if it requires a new rule?

**Comment by sunxd3** [comment]
I'll look into it 👍

File: src/flows/neuralspline.jl [hunk 1]
```diff
@@ -0,0 +1,97 @@
+##################################
+# define neural spline layer using Bijectors.jl interface
+#################################
```
Comment (Red-Portal) [inline src/flows/neuralspline.jl:3]: Seems redundant now?

File: src/flows/realnvp.jl [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    t = fnn(dim-cdims, hdims, cdims; output_activation=nothing, paramtype=paramtype)
+    mask = PartitionMask(dim, mask_idx)
+    return AffineCoupling(dim, mask, s, t)
+end
+
+function Bijectors.transform(af::AffineCoupling, x::AbstractVecOrMat)
+    # partition vector using 'af.mask::PartitionMask`
+    x₁, x₂, x₃ = partition(af.mask, x)
+    s_x₂ = af.s(x₂)
+    y₁ = x₁ .* exp.(s_x₂) .+ af.t(x₂)
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:37]: could we fuse this with `@.`? The performance of often much better this way.

File: src/flows/realnvp.jl [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+end
+
+function (af::AffineCoupling)(x::AbstractVecOrMat)
+    return transform(af, x)
+end
+
+function Bijectors.with_logabsdet_jacobian(af::AffineCoupling, x::AbstractVector)
+    x_1, x_2, x_3 = Bijectors.partition(af.mask, x)
+    s_x2 = af.s(x_2)
+    y_1 = exp.(s_x2) .* x_1 .+ af.t(x_2)
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:48]: Same as above

File: src/flows/realnvp.jl [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    logjac = sum(s_x2) # this is a scalar
+    return combine(af.mask, y_1, x_2, x_3), logjac
+end
+
+function Bijectors.with_logabsdet_jacobian(af::AffineCoupling, x::AbstractMatrix)
+    x_1, x_2, x_3 = Bijectors.partition(af.mask, x)
+    s_x2 = af.s(x_2)
+    y_1 = exp.(s_x2) .* x_1 .+ af.t(x_2)
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:56]: Same as above

File: src/flows/realnvp.jl [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+function Bijectors.with_logabsdet_jacobian(
+    iaf::Inverse{<:AffineCoupling}, y::AbstractVector
+)
+    af = iaf.orig
+    # partition vector using `af.mask::PartitionMask`
+    y_1, y_2, y_3 = partition(af.mask, y)
+    # inverse transformation
+    s_y2 = af.s(y_2)
+    x_1 = (y_1 .- af.t(y_2)) .* exp.(-s_y2)
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:70]: Same as above

File: src/flows/realnvp.jl [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+function Bijectors.with_logabsdet_jacobian(
+    iaf::Inverse{<:AffineCoupling}, y::AbstractMatrix
+)
+    af = iaf.orig
+    # partition vector using `af.mask::PartitionMask`
+    y_1, y_2, y_3 = partition(af.mask, y)
+    # inverse transformation
+    s_y2 = af.s(y_2)
+    x_1 = (y_1 .- af.t(y_2)) .* exp.(-s_y2)
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:83]: Same as above

**Review by Red-Portal — CHANGES_REQUESTED** [review]
Hi! I have some minor suggestions

File: src/flows/neuralspline.jl (hunk 1 above)
Comment (zuhengxu) [reply src/flows/neuralspline.jl:3]: yeah, will replace this with proper doc once NSF works properly (automooncake is failing on this still).

File: src/flows/realnvp.jl (hunk 2 above)
Comment (Red-Portal) [reply src/flows/realnvp.jl:37]: I don't understand why, but it appears that the compiler can't fuse these broadcast statements in a lot of cases. You can run some simple benchmarks to see if `@.` makes a difference, but in general, it seems like it does. I would also like to know why :shrug:

Comment (zuhengxu) [reply src/flows/realnvp.jl:37]: Just did a quick benchmark, seems like there is not much difference (2.212 us vs 2.18 us). The computation bottleneck for this line is the `af.t(x2)`, where `t` is a NN and already handles the batched operation.

**Comment by zuhengxu** [comment]
> It appears that the remaining issues with Mooncake are minor, likely due to a lack of a specific rule.

@yebai @sunxd3 Thanks for looking at this! I just fixed the remaining error with the nsf example; it was caused by the type instability of the nsf constructor instead of missing rules by Mooncake.jl. 

I should finish this PR soon---just need to benchmark a new NSF implementation in terms of the performance, and then update the doc.

**Comment by yebai** [comment]
>  it was caused by the type instability of the nsf constructor instead of missing rules by Mooncake.jl.

Mooncake supports type unstable code so I'm curious about why it didn't work here.

**Comment by zuhengxu** [comment]
> Mooncake supports type unstable code so I'm curious about why it didn't work here.

After further looking into this, the problem was that the type instability errored the `logpdf` evaluation of the target distribution, which is an upstream error. So there was no issue with Mooncake.

File: src/flows/neuralspline.jl
[the remark this answers is not in the document]
Comment (zuhengxu) [reply src/flows/neuralspline.jl:1]: Ah thanks! This makes sense. I think it's okay to have NSF to only support Zygote for now.  I'll go update the doc and merge the PR.

File: src/flows/utils.jl [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    input_dim::Int,
+    hidden_dims::AbstractVector{Int},
+    output_dim::Int;
+    inlayer_activation=Flux.leakyrelu,
+    output_activation=nothing,
+    paramtype::Type{T} = Float64,
+) where {T<:AbstractFloat}
+    # Create a chain of dense layers
+    # First layer
+    layers = Any[Flux.Dense(input_dim, hidden_dims[1], inlayer_activation)]
```
Comment (Red-Portal) [inline src/flows/utils.jl:80]: Aren't the types all `Flux.Dense`?

File: test/flow.jl [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            # Compute ELBO
+            batchsize = 64
+            elbo_value = elbo(Random.default_rng(), flow, logp, batchsize)
+            elbo_batch_value = elbo_batch(Random.default_rng(), flow, logp, batchsize)
+
+            # test elbo_value is not NaN and not Inf
+            @test !isnan(elbo_value)
+            @test !isinf(elbo_value)
+            @test !isnan(elbo_batch_value)
+            @test !isinf(elbo_batch_value)
```
Comment (Red-Portal) [inline test/flow.jl:60]: Checking `isfinite` handles both `NaN`s and `Inf`s.

File: test/flow.jl [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
```
Comment (Red-Portal) [inline test/flow.jl:127]: Same as bove.

File: test/flow.jl [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            μ = randn(T, dim)
+            Σ = Diagonal(rand(T, dim) .+ T(1e-3))
+            target = MvNormal(μ, Σ)
+            logp(z) = logpdf(target, z)
+
+            # Define a simple log-likelihood function
+            logp(z) = logpdf(q₀, z)
+
+            # Compute ELBO
+            batchsize = 64
```
Comment (Red-Portal) [inline test/flow.jl:52]: I have noticed that, for batch operations, `batchsize = 1` is an edge case that is easy to mess up. I recommend testing for both 1 and 64.

File: test/flow.jl [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
```
Comment (Red-Portal) [inline test/flow.jl:119]: Same as above

File: test/ad.jl [hunk 12]
```diff
@@ -73,3 +73,98 @@ end
...
+            μ = 10 * ones(T, 2)
+            Σ = Diagonal(4 * ones(T, 2))
+            target = MvNormal(μ, Σ)
+            logp(z) = logpdf(target, z)
+
+            # necessary for Zygote/mooncake to differentiate through the flow
+            # prevent updating params of q0
+            @leaf MvNormal
+            q₀ = MvNormal(zeros(T, 2), ones(T, 2))
+            flow = nsf(q₀, [8, 8], 10, 5one(T), 3; paramtype=T)
```
Comment (Red-Portal) [inline test/ad.jl:147]: Is the `5` intended?

File: src/flows/utils.jl [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+"""
+    mlp3(input_dim::Int, hidden_dims::Int, output_dim::Int; activation=Flux.leakyrelu)
+
+A simple wrapper for a 3 layer dense MLP
+"""
+function mlp3(
+    input_dim::Int, 
+    hidden_dims::Int, 
+    output_dim::Int; 
+    activation=Flux.leakyrelu,
```
Comment (Red-Portal) [inline src/flows/utils.jl:36]: Just curious. Do leaky ReLUs generally work better than `tanh` for flows?

File: src/flows/realnvp.jl [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+[^LJS2017]: Dinh, L., Sohl-Dickstein, J. and Bengio, S. (2017).  Density estimation using Real NVP. ICLR.
+"""
+struct AffineCoupling <: Bijectors.Bijector
+    dim::Int
+    mask::Bijectors.PartitionMask
+    s::Flux.Chain
+    t::Flux.Chain
+end
+
+# let params track field s and t
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:40]: Probably redundant?

File: src/flows/realnvp.jl [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+@functor AffineCoupling (s, t)
+
+function AffineCoupling(
+    dim::Int,                       # dimension of the problem
+    hdims::AbstractVector{Int},     # dimension of hidden units for s and t
+    mask_idx::AbstractVector{Int},       # index of dimensione that one wants to apply transformations on
+    paramtype::Type{T}
```
Comment (Red-Portal) [inline src/flows/realnvp.jl:47]: I think the comments here are unnecessary given the docstrings above.

File: src/flows/neuralspline.jl [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+- `paramtype::Type{T} = Float64`: parameter element type.
+
+Returns
+- `Bijectors.TransformedDistribution` representing the NSF flow.
+
+Notes:
+- Under the hood, `nsf` relies on the rational quadratic spline function implememented in 
+`MonotonicSplines.jl` for performance reasons.  `MonotonicSplines.jl` uses 
+`KernelAbstractions.jl` to support batched operations. 
+Because of this, so far `nsf` only supports `Zygote` as the AD type.
```
Comment (Red-Portal) [inline src/flows/neuralspline.jl:212]: There is a nice feature in `Documenter` that highlights important information: See the `!!!` series [here](https://documenter.juliadocs.org/stable/showcase/#Admonitions).

File: src/flows/neuralspline.jl [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+[^DBMP2019]: Durkan, C., Bekasov, A., Murray, I. and Papamarkou, T. (2019). Neural Spline Flows. *NeurIPS.*
+"""
+struct NeuralSplineCoupling{T,A<:Flux.Chain} <: Bijectors.Bijector
+    dim::Int                        # dimension of input
+    K::Int                          # number of knots
+    n_dims_transferred::Int         # number of dimensions that are transformed
+    B::T                            # bound of the knots
+    nn::A                           # networks that parmaterize the knots and derivatives
+    mask::Bijectors.PartitionMask
+end
```
Comment (Red-Portal) [inline src/flows/neuralspline.jl:42]: `DocStringExtensions` supports docstrings for the struct fields. See [here](https://discourse.julialang.org/t/documenting-elements-of-a-struct/64769/11?u=red-portal)

File: src/flows/neuralspline.jl [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+
+function NeuralSplineCoupling(
+    dim::T1,                         # dimension of input
+    hdims::AbstractVector{T1},       # dimension of hidden units for s and t
+    K::T1,                           # number of knots
+    B::T2,                           # bound of the knots
+    mask_idx::AbstractVector{T1}, # index of dimensione that one wants to apply transformations on
+    paramtype::Type{T2},             # type of the parameters, e.g., Float64 or Float32
+) where {T1<:Int,T2<:AbstractFloat}
```
Comment (Red-Portal) [inline src/flows/neuralspline.jl:51]: The comments seem unnecessary given the docstrings?

File: src/flows/neuralspline.jl [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+    num_of_transformed_dims = length(mask_idx)
+    input_dims = dim - num_of_transformed_dims
+    
+    # output dim of the NN
```
Comment (Red-Portal) [inline src/flows/neuralspline.jl:55]: I think the variable name is pretty self-explanatory.

File: example/Project.toml [hunk 20]
```diff
@@ -3,9 +3,10 @@ ADTypes = "47edcb42-4c32-4615-8424-f2b9edc5f35b"
 BenchmarkTools = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
 Bijectors = "76274a88-744f-5084-9051-94815aaf08c4"
 DiffResults = "163ba53b-c6d8-5494-b064-1a9d43ac40c5"
+DifferentiationInterface = "a0c0ee7d-e4b9-4e03-894e-1c5f64a51d63"
 Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
 DocStringExtensions = "ffbed154-4ef7-542d-bbb7-c09d3a79fcae"
-Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
```
Comment (Red-Portal) [inline example/Project.toml:8]: Seems like one of the examples uses `Zygote`?

**Review by Red-Portal — CHANGES_REQUESTED** [review]
I only have minor suggestions. Feel free to take a look and apply them if you agree. Otherwise, looks good to me.

File: docs/src/PlanarFlow.md [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+```julia
+using Bijectors
+using Functors # for @leaf
+
+function create_planar_flow(n_layers::Int, q₀)
+    d = length(q₀)
+    Ls = [PlanarLayer(d) for _ in 1:n_layers]
+    ts = reduce(∘, Ls)  # alternatively: FunctionChains.fchain(Ls)
+    return transformed(q₀, ts)
+end
```
Comment (Red-Portal) [inline docs/src/PlanarFlow.md:45]: I think we discussed this at some point, but I feel this function could also be made part of the library. (With a function name more consistent with the other constructors.) That way, the user won't have to load `Bijectors`. What do you think?

File: src/flows/utils.jl (hunk 13 above)
Comment (zuhengxu) [reply src/flows/utils.jl:36]: In this context of NF, my experience is that `tanh`, because that it restricts the output to [0, 1], you generally need more layers; the benefit is that it's generally more stable in training. In most of the NF literature (to my knowledge), `tanh` mostly gets used in the scaling NN of affine coupling to avoid overflow; `leakyrelu` is the default activation for NN in other places.

File: example/Project.toml (hunk 20 above)
Comment (zuhengxu) [reply example/Project.toml:8]: I included `Zygote` explicitly. All the examples (`demo_...`) runs properly as part of CI.

**Review by sunxd3 — COMMENTED** [review]
couple of tiny things, very happy to do another round of review

File: test/ad.jl (hunk 12 above)
Comment (zuhengxu) [reply test/ad.jl:147]: no, but 5 is a default value that's often used in NSF in literature

File: src/flows/utils.jl [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+using Flux
+
+"""
+    create_flow(layers, q0)
+
+Construct a normalizing flow by composing the provided bijector layers and
+attaching them to the base distribution `q0`.
+
+- `layers`: an iterable of `Bijectors.Bijector` objects that are composed in order
+  (left-to-right) via function composition.
```
[the remark this answers is not in the document]
Comment (zuhengxu) [reply src/flows/utils.jl:11]: Ah I should definitely make this explicit; I myself often forget as well.

> (is my order in the suggestions even correct?)

It will be the other way around---the flow will be `l3∘l2∘l1(q0)`.

**Comment by zuhengxu** [comment]
Thank you @sunxd3 @Red-Portal for the review! I made the corresponding updates and let me know if the current version looks good to you!

File: docs/src/example.md [hunk 23]
```diff
@@ -1,122 +0,0 @@
...
-scatter!(samples_untrained[1, :], samples_untrained[2, :]; label="Untrained Flow", color=:red, markersize=2, alpha=0.5)
-scatter!(samples_trained[1, :], samples_trained[2, :]; label="Trained Flow", color=:green, markersize=2, alpha=0.5)
-plot!(title = "Comparison of Trained and Untrained Flow", xlabel = "X", ylabel= "Y", legend=:topleft) 
-```
-![compare](comparison.png)
-
-
-## Reference 
-
-- Rezende, D. and Mohamed, S., 2015. *Variational inference with normalizing flows*. International Conference on Machine Learning
```
Comment (sunxd3) [inline docs/src/example.md:122]: think we can just remove this file?

**Review by sunxd3 — COMMENTED** [review]
another couple of tiny things, nothing major beyond these

File: src/flows/utils.jl [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+
+"""
+    create_flow(layers, q0)
+
+Construct a normalizing flow by composing the provided bijector layers and
+attaching them to the base distribution `q0`.
+
+- `layers`: an iterable of `Bijectors.Bijector` objects that are composed in order
+  (left-to-right) via function composition 
+(for instance, if `layers = [l1, l2, l3]`, the flow will be `l3∘l2∘l1(q0)`).
```
[the remark this answers is not in the document]
Comment (zuhengxu) [reply src/flows/utils.jl:12]: I made a mistake there; you were correct, `reduce(∘, Ls) ` returns `l1 ∘ l2 ∘ l3`, see the following:
```julia
julia> f1 = exp
exp (generic function with 13 methods)

julia> f2(x) = x^2
f2 (generic function with 1 method)

julia> f3(x) = x+1
f3 (generic function with 1 method)

julia> Ls = [f1, f2, f3]
3-element Vector{Function}:
 exp (generic function with 13 methods)
 f2 (generic function with 1 method)
 f3 (generic function with 1 method)

julia> ff = reduce(∘, Ls)
exp ∘ f2 ∘ f3
```

**Review by Red-Portal — CHANGES_REQUESTED** [review]
I only have a few minor comments that should be quick to handle!

**Comment by zuhengxu** [comment]
@Red-Portal @sunxd3 Let me know if I can hit the big green button! Thanks for the quick feedback.

**Comment by Red-Portal** [comment]
Alright looks good to me now!
